lib/rule_engine/ast.py

```diff
@@ -377,14 +381,38 @@ def __op_arithmetic(self, op, thing):
 		_assert_is_numeric(right_value)
 		return op(left_value, right_value)
 
-	_op_add  = functools.partialmethod(__op_arithmetic, operator.add)
 	_op_sub  = functools.partialmethod(__op_arithmetic, operator.sub)
 	_op_fdiv = functools.partialmethod(__op_arithmetic, operator.floordiv)
 	_op_tdiv = functools.partialmethod(__op_arithmetic, operator.truediv)
 	_op_mod  = functools.partialmethod(__op_arithmetic, operator.mod)
 	_op_mul  = functools.partialmethod(__op_arithmetic, operator.mul)
 	_op_pow  = functools.partialmethod(__op_arithmetic, operator.pow)
 
+class AddExpression(LeftOperatorRightExpressionBase):
+	"""A class for representing arithmetic expressions from the grammar text such as addition and subtraction."""
+	compatible_types = (DataType.FLOAT,DataType.STRING)
+	result_type =  DataType.UNDEFINED
+
+	def __init__(self, *args, **kwargs):
+		super(AddExpression, self).__init__(*args, **kwargs)
+		if self.left.result_type != DataType.UNDEFINED and self.right.result_type != DataType.UNDEFINED:
+			if self.left.result_type != self.right.result_type:
+				raise errors.EvaluationError('data type mismatch')
+
+	def __op_add(self, op, thing):
+		left_value = self.left.evaluate(thing)
+		right_value = self.right.evaluate(thing)
+		if isinstance(left_value,str) or isinstance(right_value, str):
+			_assert_is_string(left_value)
+			_assert_is_string(right_value)
+		else:
+			_assert_is_numeric(left_value)
+			_assert_is_numeric(right_value)
```

Comment:
The `_assert_is*` functions take `*args` so they only need to be called once to check both values.

```suggestion
		if isinstance(left_value, str) or isinstance(right_value, str):
			_assert_is_string(left_value, right_value)
		else:
			_assert_is_numeric(left_value, right_value)
```